doc/sphinx/Pacemaker_Development/c.rst

```diff
@@ -371,72 +442,163 @@ Control Statements (if, else, while, for, switch)
+    *     function to be used by more callers.
+    */
+   int
+   my_func1(const char *s)
+   {
+       return 0;
+   }
+
+   /* Functions with no arguments must explicitly list them as void,
+    * for compatibility with strict compilers
+    */
+   int
+   my_func2(void)
+   {
+       return 0;
+   }
 
+   /*
+    * (1) For functions with enough arguments that they must break to the next
+    *     line, align arguments with the first argument.
+    * (2) When a function argument is a function itself, use the pointer form.
+    * (3) Declare functions and file-global variables as ``static`` whenever
+    *     appropriate. This gains a slight efficiency in shared libraries, and
+    *     helps the reader know that it is not used outside the one file.
+    */
+   static int
+   my_func3(int bar, const char *a, const char *b, const char *c,
+            void (*callback)())
+   {
+       return 0;
+   }
 
 
+Return Values
+_____________
 
-.. index::
-   pair: C; struct
-   pair: C; enum
+Functions that need to indicate success or failure should follow one of the
+following guidelines. More details, including functions for using them in user
+messages and converting from one to another, can be found in
+``include/crm/common/results.h``.
 
-New Struct and Enum Members
-###########################
+* A **standard Pacemaker return code** is one of the ``pcmk_rc_*`` enum values
+  or a system errno code, as an ``int``.
 
-In the public APIs, always add new ``struct`` members to the end of the
-``struct``. This allows us to maintain backward API/ABI compatibility (as long
-as the application being linked allocates structs via API functions).
+* ``crm_exit_t`` (the ``CRM_EX_*`` enum values) is a system-independent code
+  suitable for the exit status of a process, or for interchange between nodes.
 
-This generally applies to ``enum`` values as well, as the compiler will define
-``enum`` values to 0, 1, etc., in the order given, so inserting a value in the
-middle will change the numerical values of all later values, making them
-backward-incompatible. However, if enum numerical values are explicitly
-specified rather than left to the compiler, new values can be added anywhere.
+* Other special-purpose status codes exist, such as ``enum ocf_exitcode`` for
+  the possible exit statuses of OCF resource agents (along with some
+  Pacemaker-specific extensions). It is usually obvious when the context calls
+  for such.
 
+* Some older Pacemaker APIs use the now-deprecated "legacy" return values of
+  ``pcmk_ok`` or the positive or negative value of one of the ``pcmk_err_*``
+  constants or system errno codes.
 
+* Functions registered with external libraries (as callbacks for example)
+  should use the appropriate signature defined by those libraries, rather than
+  follow Pacemaker guidelines.
 
+Of course, functions may have return values that aren't success/failure
+indicators, such as a pointer, integer count, or bool.
```

Comment:
It'd be nice to see an explicit mention of the new `pcmk_rc_` return codes.

Reply:
What level of detail beyond the current bullet point and pointing to the header?

Reply:
Hah, whoops.  I missed that up there.  Looks good.